src/deploy/functions/iambinder.ts

```diff
@@ -0,0 +1,298 @@
+  return s.startsWith("roles/");
+}
+
+interface IamBinding extends Omit<iam.Binding, "members" | "role"> {
+  role: Role;
+  members: Set<string>;
+}
+
+/* @internal */
+export class IamBindings {
+  additions: Record<Role, Record<string, IamBinding>>; // role -> condition -> bindings
+
+  constructor() {
+    this.additions = {};
+  }
+
+  add(role: Role, members: string[], condition?: iam.Binding["condition"]) {
+    const conditions = this.additions[role] || {};
+    const conditionKey = condition ? IamBindings.getConditionKey(condition) : "";
+    const binding = conditions[conditionKey] || { role, members: new Set() };
+    if (condition) {
+      binding.condition = { ...condition };
+    }
+
+    for (const member of members) {
+      binding.members.add(member);
+    }
+
+    conditions[conditionKey] = binding;
+    this.additions[role] = conditions;
+  }
+
+  diff(base: IamBindings): IamBindings {
+    const diff = this.clone();
+    for (const { role, members, condition } of base.allBindings()) {
+      const conditionKey = condition ? IamBindings.getConditionKey(condition) : "";
+      if (diff.additions[role]?.[conditionKey]) {
+        for (const member of members) {
+          diff.additions[role][conditionKey].members.delete(member);
+        }
+        if (diff.additions[role][conditionKey].members.size === 0) {
+          delete diff.additions[role][conditionKey];
+        }
+        if (Object.keys(diff.additions[role]).length === 0) {
+          delete diff.additions[role];
+        }
+      }
+    }
+    return diff;
+  }
+
+  merge(other: IamBindings): IamBindings {
+    const merged = new IamBindings();
+    for (const bindings of [this, other]) {
+      for (const binding of bindings.allBindings()) {
+        merged.add(binding.role, Array.from(binding.members), binding.condition);
+      }
+    }
+    return merged;
+  }
+
+  clone(): IamBindings {
+    const copied = new IamBindings();
+    for (const binding of this.allBindings()) {
+      copied.add(binding.role, Array.from(binding.members), binding.condition);
+    }
+    return copied;
+  }
+
+  asIamBindings(): iam.Binding[] {
+    const iamBindings: iam.Binding[] = [];
+    for (const binding of this.allBindings()) {
+      const copy: iam.Binding = { role: binding.role, members: Array.from(binding.members) };
+      if (binding.condition) {
+        copy.condition = { ...binding.condition };
+      }
+      iamBindings.push(copy);
+    }
+    return iamBindings;
+  }
+
+  private *allBindings(): Generator<IamBinding> {
+    for (const role of Object.keys(this.additions) as Role[]) {
+      for (const conditionKey of Object.keys(this.additions[role])) {
+        yield this.additions[role][conditionKey];
+      }
+    }
+  }
+
+  private static getConditionKey(condition: iam.Binding["condition"]): string {
+    return createHash("md5").update(JSON.stringify(condition)).digest("base64");
+  }
+
+  static fromIamBindings(bindings: iam.Binding[]): IamBindings {
```

Comment:
I think this function name should be singular - `fromIamBinding` - since we're converting from the `iam.Binding` array, right?

Reply:
I'm expecting bindings from `policy.bindings` which is plural. I _think_ I'd refer to an array of `iam.Binding` in plural form - is that what we do in other places 🤔 ?

Reply:
An array of `foo` is a `foos`.